clang/lib/Sema/SemaChecking.cpp

```diff
@@ -8454,26 +8454,43 @@ static bool IsInfOrNanFunction(StringRef calleeName, MathCheck Check) {
   llvm_unreachable("unknown MathCheck");
 }
 
+static bool IsInfinityFunction(const FunctionDecl *FDecl) {
+  if (FDecl->getName() != "infinity")
+    return false;
+
+  if (const CXXMethodDecl *MDecl = dyn_cast<CXXMethodDecl>(FDecl)) {
+    const CXXRecordDecl *RDecl = MDecl->getParent();
+    if (RDecl->getName() != "numeric_limits")
+      return false;
+
+    if (const NamespaceDecl *NSDecl =
+            dyn_cast<NamespaceDecl>(RDecl->getDeclContext()))
+      return NSDecl->isStdNamespace();
+  }
+
+  return false;
+}
+
 void Sema::CheckInfNaNFunction(const CallExpr *Call,
                                const FunctionDecl *FDecl) {
+  if (!FDecl->getIdentifier())
```

Comment:
Wouldn't something like this work?
```suggestion
  if (!FDecl->getIdentifier())
  return;
FPOptions FPO = Call->getFPFeaturesInEffect(getLangOpts());
bool IsNaNOrIsUnordered =
    IsStdFunction(FDecl, "isnan") || IsStdFunction(FDecl, "isunordered");
bool IsSpecialNaN =
    IsInfOrNanFunction(FDecl->getName(), MathCheck::NaN);
bool IsInfOrIsFinite =
    IsStdFunction(FDecl, "isinf") || IsStdFunction(FDecl, "isfinite");
bool IsInfinityOrIsSpecialInf =
    IsInfOrNanFunction(FDecl->getName(), MathCheck::Inf);

if ((IsNaNOrIsUnordered || IsSpecialNaN) && FPO.getNoHonorNaNs()) {
  Diag(Call->getBeginLoc(), diag::warn_fp_nan_inf_when_disabled)
      << 1 << 0 << Call->getSourceRange();
  return;
}
if ((IsInfOrIsFinite || IsInfinityFunction(FDecl) ||
     IsInfinityOrIsSpecialInf) &&
    FPO.getNoHonorInfs()) {
  Diag(Call->getBeginLoc(), diag::warn_fp_nan_inf_when_disabled)
      << 0 << 0 << Call->getSourceRange();
}
```

Reply:
The problem with this is in case of `FPO.getNoHonorNaNs() == false` we performed un needed 3 function call, and in case of it's true we performed 3 more un needed too `IsInfOrIsFinite`, `IsInfinityOrIsSpecialInf`

```call
bool IsNaNOrIsUnordered =
    IsStdFunction(FDecl, "isnan") || IsStdFunction(FDecl, "isunordered");
bool IsSpecialNaN =
    IsInfOrNanFunction(FDecl->getName(), MathCheck::NaN);
    
if ((IsNaNOrIsUnordered || IsSpecialNaN) && false) { <--- Condition will evaluated to false

}
```

Reply:
Oh! right, That makes sense!

Reply:
Thank you